monitor/internal/pod/controller.go

```diff
@@ -0,0 +1,339 @@
+func addController(mgr manager.Manager, r *ReconcilePod, eventsCh <-chan event.GenericEvent) error {
+	// Create a new controller
+	c, err := controller.New("trireme-pod-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// we use this mapper in both of our event sources
+	mapper := &WatchPodMapper{
+		client:         mgr.GetClient(),
+		nodeName:       r.nodeName,
+		enableHostPods: r.enableHostPods,
+	}
+
+	// use the our watch pod mapper which filters pods before we reconcile
+	if err := c.Watch(
+		&source.Kind{Type: &corev1.Pod{}},
+		&handler.EnqueueRequestsFromMapFunc{ToRequests: mapper},
+	); err != nil {
+		return err
+	}
+
+	// we pass in a custom channel for events generated by resync
+	if err := c.Watch(
+		&source.Channel{Source: eventsCh},
+		&handler.EnqueueRequestsFromMapFunc{ToRequests: mapper},
+	); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcilePod{}
+
+// ReconcilePod reconciles a Pod object
+type ReconcilePod struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client            client.Client
+	scheme            *runtime.Scheme
+	recorder          record.EventRecorder
+	handler           *config.ProcessorConfig
+	metadataExtractor extractors.PodMetadataExtractor
+	netclsProgrammer  extractors.PodNetclsProgrammer
+	nodeName          string
+	enableHostPods    bool
+
+	metadataExtractTimeout time.Duration
+	handlePUEventTimeout   time.Duration
+	netclsProgramTimeout   time.Duration
+}
+
+// Reconcile reads that state of the cluster for a pod object
+func (r *ReconcilePod) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	ctx := context.Background()
+	puID := request.NamespacedName.String()
+
+	// Fetch the corresponding pod object.
+	pod := &corev1.Pod{}
+	if err := r.client.Get(ctx, request.NamespacedName, pod); err != nil {
+		if errors.IsNotFound(err) {
+			zap.L().Debug("Pod IsNotFound", zap.String("puID", puID))
+			handlePUCtx, handlePUCancel := context.WithTimeout(ctx, r.handlePUEventTimeout)
+			defer handlePUCancel()
+
+			// try to call stop, but don't fail if that errors
+			err := r.handler.Policy.HandlePUEvent(
+				handlePUCtx,
+				puID,
+				common.EventStop,
+				policy.NewPURuntimeWithDefaults(),
+			)
+			if err != nil {
+				zap.L().Warn("failed to handle stop event during destroy", zap.String("puID", puID), zap.Error(err))
+			}
+
+			// call destroy regardless if stop succeeded
+			if err := r.handler.Policy.HandlePUEvent(
+				handlePUCtx,
+				puID,
+				common.EventDestroy,
+				policy.NewPURuntimeWithDefaults(),
+			); err != nil {
+				zap.L().Error("failed to handle destroy event", zap.String("puID", puID), zap.Error(err))
+				return reconcile.Result{}, ErrHandlePUDestroyEventFailed
+			}
+			return reconcile.Result{}, nil
```

Comment:
Can you add a note here, that we should just need one of the two events in engine and we need to fix that there.

Reply:
will do